test/include/util/storage_test_util.h

```diff
@@ -268,4 +275,173 @@ struct StorageTestUtil {
     return storage::BlockLayout(attr_sizes);
   }
 };
+
+class RandomSqlTableTestObject {
+ public:
+  RandomSqlTableTestObject() = default;
+  ~RandomSqlTableTestObject() {
+    delete pri_;
+    delete pr_map_;
+    delete schema_;
+    delete layout_;
+    delete table_;
+  }
+
+  /**
+   * Generate random columns, and add them sequencially to the internal list.
+   * @param num_cols number of columns to add.
+   * @param varlen_allowed allow VARCHAR to be used in columns.
+   * @param generator of random numbers.
+   */
+  template <class Random>
+  void GenerateRandomColumns(int num_cols, bool varlen_allowed, Random *generator) {
+    std::string prefix = "col_";
+    std::vector<type::TypeId> types = DataTypeAll(varlen_allowed);
+    for (int i = 0; i < num_cols; i++) {
+      type::TypeId type = *RandomTestUtil::UniformRandomElement(&types, generator);
+      DefineColumn(prefix + std::to_string(uint8_t(type)), type, true, catalog::col_oid_t(i));
+    }
+  }
+
+  /**
+   * Append a column definition to the internal list. The list will be
+   * used when creating the SqlTable.
+   * @param name of the column
+   * @param type of the column
+   * @param nullable
+   * @param oid for the column
+   */
+  void DefineColumn(std::string name, type::TypeId type, bool nullable, catalog::col_oid_t oid) {
+    cols_.emplace_back(name, type, nullable, oid);
+  }
+
+  /**
+   * Create the SQL table.
+   */
+  void Create() {
+    schema_ = new catalog::Schema(cols_);
+    table_ = new storage::SqlTable(&block_store_, *schema_, table_oid_);
+    layout_ = new storage::BlockLayout(storage::StorageUtil::BlockLayoutFromSchema(*schema_).first);
+
+    for (const auto &c : cols_) {
+      col_oids_.emplace_back(c.GetOid());
+    }
+
+    // save information needed for (later) reading and writing
+    auto row_pair = table_->InitializerForProjectedRow(col_oids_);
+    pri_ = new storage::ProjectedRowInitializer(std::get<0>(row_pair));
+    pr_map_ = new storage::ProjectionMap(std::get<1>(row_pair));
+  }
+
+  template <class Random>
+  void InsertRandomRow(transaction::TransactionContext *txn, const double null_bias, Random *generator) {
+    std::bernoulli_distribution coin(1 - null_bias);
+    std::uniform_int_distribution<uint32_t> varlen_size(1, 2 * storage::VarlenEntry::InlineThreshold());
+    std::uniform_int_distribution<uint8_t> char_dist(0, UINT8_MAX);
+    std::uniform_int_distribution<int32_t> int_dist(INT32_MIN, INT32_MAX);
+
+    auto insert_buffer = common::AllocationUtil::AllocateAligned(pri_->ProjectedRowSize());
+    auto insert = pri_->InitializeRow(insert_buffer);
+
+    for (int i = 0; i < static_cast<int>(cols_.size()); i++) {
+      if (coin(*generator)) {  // not null
+        uint16_t offset = pr_map_->at(col_oids_[i]);
+        insert->SetNotNull(offset);
```

Comment:
You don't need to SetNotNull AccessForceNotNull you do below already does that

Reply:
resolved.